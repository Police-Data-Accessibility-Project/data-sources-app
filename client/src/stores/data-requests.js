import { defineStore } from 'pinia';

export const useDataRequestsStore = defineStore('data-requests', {
	state: () => ({
		/** Searches performed during session. */
		cache: {},
		/** Previous route visited - useful for determining whether we are incrementing or decrementing pages in data request by id */
		previousDataRequestRoute: null,
	}),
	getters: {
		getDataRequestFromCache: (state) => (key) => {
			if (!(key in state.cache)) {
				return null;
			}
			return state.cache[key];
		},
	},

	actions: {
		clearCache() {
			this.$patch((state) => {
				state.cache = {};
			});
		},
		setDataRequestToCache(key, data, timestamp = new Date().getTime()) {
			this.$patch((state) => {
				state.cache[key] = {
					data,
					timestamp,
				};
			});
		},
		setPreviousDataRequestRoute(route) {
			this.$patch({
				previousDataSourceRoute: route,
			});
		},
	},
});
