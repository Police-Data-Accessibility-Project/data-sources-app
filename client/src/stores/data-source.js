import { defineStore } from 'pinia';

export const useDataSourceStore = defineStore('data-source', {
	state: () => ({
		/** Previous route visited - useful for determining whether we are incrementing or decrementing pages in data source by id */
		previousDataSourceRoute: null,
		/** Cache by ID -- useful for avoiding unnecessary API calls */
		cache: {},
	}),
	persist: {
		storage: sessionStorage,
	},
	getters: {
		getDataSourceFromCache: (state) => (key) => {
			if (!(key in state.cache)) {
				return null;
			}
			return state.cache[key];
		},
	},
	actions: {
		clearCache() {
			this.$patch((state) => {
				state.cache = {};
			});
		},

		setPreviousDataSourceRoute(route) {
			this.$patch({
				previousDataSourceRoute: route,
			});
		},

		setDataSourceToCache(key, data) {
			this.$patch((state) => {
				state.cache[key] = {
					data,
					timestamp: new Date().getTime(),
				};
			});
		},
	},
});
