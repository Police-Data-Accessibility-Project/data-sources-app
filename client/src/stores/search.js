import { defineStore } from 'pinia';

export const useSearchStore = defineStore('search', {
	state: () => ({
		/** Searches performed during session. */
		cache: {},
		/** Needed for `NEXT` / `BACK` functionality in data source id view */
		mostRecentSearchIds: [],
		/** Needed for `NEXT` / `BACK` functionality in data requests id view */
		mostRecentRequestIds: [],
	}),
	persist: {
		storage: sessionStorage,
		pick: ['mostRecentSearchIds'],
	},
	getters: {
		getSearchFromCache: (state) => (key) => {
			if (!(key in state.cache)) {
				return null;
			}
			return state.cache[key];
		},
	},

	actions: {
		clearCache() {
			this.$patch((state) => {
				state.cache = {};
			});
		},

		setMostRecentSearchIds(ids) {
			this.$patch({
				mostRecentSearchIds: ids,
			});
		},

		setMostRecentRequestIds(ids) {
			this.$patch({
				mostRecentRequestIds: ids,
			});
		},

		setSearchToCache(key, data, timestamp = new Date().getTime()) {
			this.$patch((state) => {
				// Use object notation for setting cache data
				state.cache[key] = {
					data,
					timestamp,
				};
			});
		},
	},
});
