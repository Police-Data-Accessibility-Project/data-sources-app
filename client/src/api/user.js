import axios from 'axios';
import { ENDPOINTS } from './constants';
import { useAuthStore } from '@/stores/auth';
import { useUserStore } from '@/stores/user';

const USER_BASE = `${import.meta.env.VITE_VUE_API_BASE_URL}/user`;
const HEADERS = {
	'Content-Type': 'application/json',
};

export async function changePassword(oldPassword, newPassword) {
	const auth = useAuthStore();
	const user = useUserStore();

	await axios.post(
		`${USER_BASE}/${user.id}/${ENDPOINTS.USER.ID.UPDATE_PASSWORD}`,
		{ old_password: oldPassword, new_password: newPassword },
		{
			headers: {
				...HEADERS,
				Authorization: `Bearer ${auth.$state.tokens.accessToken.value}`,
			},
		},
	);

	return await auth.signInWithEmail(user.email, newPassword);
}
