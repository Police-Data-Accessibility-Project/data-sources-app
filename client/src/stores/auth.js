import axios from 'axios';
import { defineStore } from 'pinia';
import parseJwt from '../util/parseJwt';
import router from '../router';
import { useUserStore } from './user';

const HEADERS = {
	headers: { 'Content-Type': 'application/json' },
};
const LOGIN_URL = `${import.meta.env.VITE_VUE_API_BASE_URL}/login`;
const REFRESH_SESSION_URL = `${import.meta.env.VITE_VUE_API_BASE_URL}/refresh-session`;

export const useAuthStore = defineStore('auth', {
	state: () => ({
		userId: null,
		accessToken: {
			value: null,
			expires: Date.now(),
		},
		returnUrl: null,
	}),
	persist: true,
	actions: {
		async login(email, password) {
			const user = useUserStore();

			try {
				const response = await axios.post(
					LOGIN_URL,
					{ email, password },
					HEADERS,
				);

				// Update user store with email
				user.$patch({ email });

				this.parseTokenAndSetData(response);
				if (this.returnUrl) router.push(this.returnUrl);
			} catch (error) {
				throw new Error(error.response?.data?.message);
			}
		},

		logout(isAuthRoute) {
			const user = useUserStore();

			this.$patch({
				userId: null,
				accessToken: { value: null, expires: Date.now() },
				returnUrl: null,
			});

			user.$patch({
				email: '',
			});
			if (isAuthRoute) router.push('/login');
		},

		async refreshAccessToken() {
			if (!this.$state.userId) return;
			try {
				const response = await axios.post(
					REFRESH_SESSION_URL,
					{ session_token: this.$state.accessToken.value },
					HEADERS,
				);
				return this.parseTokenAndSetData(response);
			} catch (error) {
				throw new Error(error.response?.data?.message);
			}
		},

		parseTokenAndSetData(response) {
			const token = response.data.data;
			const tokenParsed = parseJwt(token);

			this.$patch({
				userId: tokenParsed.sub,
				accessToken: {
					value: token,
					expires: new Date(tokenParsed.exp * 1000).getTime(),
				},
			});
		},
	},
});
