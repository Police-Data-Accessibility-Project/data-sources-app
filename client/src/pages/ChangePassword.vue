<template>
	<main v-if="success" class="pdap-flex-container">
		<h1>Success</h1>
		<p>Your password has been successfully updated</p>
	</main>
	<main v-else class="pdap-flex-container mx-auto max-w-2xl">
		<h1>Change your password</h1>
		<Form
			id="change-password"
			class="flex flex-col"
			data-test="change-password-form"
			name="change-password"
			:error="error"
			:schema="FORM_SCHEMA"
			@change="handleChangeOnError"
			@submit="onSubmit"
		>
			<Button class="max-w-full" type="submit">
				{{ loading ? 'Loading...' : 'Change password' }}
			</Button>
		</Form>
	</main>
</template>

<script setup>
import { Button, Form } from 'pdap-design-system';
import { useUserStore } from '../stores/user';
import { ref } from 'vue';

// Constants
const FORM_SCHEMA = [
	{
		'data-test': 'password',
		id: 'password',
		name: 'password',
		label: 'Password',
		type: 'password',
		placeholder: 'Your updated password',
		value: '',
		validators: {
			password: {
				message: 'Please provide your password',
				value: true,
			},
		},
	},
	{
		'data-test': 'confirm-password',
		id: 'confirmPassword',
		name: 'confirmPassword',
		label: 'Confirm Password',
		type: 'password',
		placeholder: 'Confirm your updated password',
		value: '',
		validators: {
			password: {
				message: 'Please confirm your password',
				value: true,
			},
		},
	},
];

// Stores
const user = useUserStore();

// Reactive vars
const error = ref(undefined);
const loading = ref(false);
const success = ref(false);

// Functions
// Handlers
/**
 * Handles clearing pw-match form errors on change if they exist
 */
function handleChangeOnError(formValues) {
	if (error.value && formValues.password !== formValues.confirmPassword) {
		handlePasswordValidation(formValues);
	}
}

/**
 * Validates that passwords match
 * @returns {boolean} `false` if passwords do not match, `true` if they do
 */
function handlePasswordValidation(formValues) {
	if (formValues.password !== formValues.confirmPassword) {
		if (!error.value) {
			error.value = 'Passwords do not match, please try again.';
		}
		return false;
	} else {
		error.value = undefined;
		return true;
	}
}

/**
 * Updates user's password
 */
async function onSubmit(formValues) {
	if (!handlePasswordValidation(formValues)) return;

	try {
		loading.value = true;
		const { password } = formValues;
		await user.changePassword(user.email, password);

		success.value = true;
	} catch (err) {
		error.value = err.message;
	} finally {
		loading.value = false;
	}
}
</script>
