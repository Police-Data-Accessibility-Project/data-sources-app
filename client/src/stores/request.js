import { useAuthStore } from '@/stores/auth';
import axios from 'axios';
import { defineStore } from 'pinia';

// Constants
const HEADERS_BASE = {
	authorization: `Basic ${import.meta.env.VITE_ADMIN_API_KEY}`,
	'Content-Type': 'application/json',
};
const REQUESTS_URL = `${import.meta.env.VITE_VUE_API_BASE_URL}/data-requests`;

export const useRequestStore = defineStore('request', {
	state: () => ({}),
	// persist: {
	// 	storage: sessionStorage,
	// },
	actions: {
		async createRequest(data) {
			const { tokens } = useAuthStore();
			const response = await axios.post(REQUESTS_URL, data, {
				headers: {
					...HEADERS_BASE,
					authorization: `Bearer ${tokens.accessToken.value}`,
				},
			});

			return response.data;
		},
	},
});
