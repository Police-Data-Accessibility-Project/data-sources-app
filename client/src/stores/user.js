import axios from 'axios';
import { defineStore } from 'pinia';
import { useAuthStore } from './auth';

const HEADERS = {
	headers: { 'Content-Type': 'application/json' },
};
const SIGNUP_URL = `${import.meta.env.VITE_VUE_API_BASE_URL}/user`;
const CHANGE_PASSWORD_URL = `${import.meta.env.VITE_VUE_API_BASE_URL}/user`;
const REQUEST_PASSWORD_RESET_URL = `${import.meta.env.VITE_VUE_API_BASE_URL}/request-reset-password`;
const PASSWORD_RESET_URL = `${import.meta.env.VITE_VUE_API_BASE_URL}/reset-password`;

export const useUserStore = defineStore('user', {
	state: () => ({
		email: '',
	}),
	persist: true,
	actions: {
		async signup(email, password) {
			const auth = useAuthStore();

			try {
				await axios.post(SIGNUP_URL, { email, password }, HEADERS);
				// Update store with email
				this.$patch({ email });
				// Log users in after signup and return that response
				return await auth.login(email, password);
			} catch (error) {
				throw new Error(error.response?.data?.message);
			}
		},

		async changePassword(email, password) {
			const auth = useAuthStore();
			try {
				await axios.put(
					CHANGE_PASSWORD_URL,
					{ email, password },
					{
						headers: {
							...HEADERS.headers,
							Authorization: `Bearer ${auth.accessToken.value}`,
						},
					},
				);
				return await auth.login(email, password);
			} catch (error) {
				throw new Error(error.response?.data?.message);
			}
		},

		async requestPasswordReset(email) {
			try {
				await axios.post(REQUEST_PASSWORD_RESET_URL, { email }, HEADERS);
			} catch (error) {
				throw new Error(error.response?.data?.message);
			}
		},

		async resetPassword(password, token) {
			try {
				await axios.post(`${PASSWORD_RESET_URL}`, { password, token }, HEADERS);
			} catch (error) {
				throw new Error(error.response?.data?.message);
			}
		},
	},
});
