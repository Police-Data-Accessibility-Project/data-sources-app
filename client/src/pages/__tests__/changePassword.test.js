import { flushPromises, mount } from '@vue/test-utils';
import { describe, expect, beforeEach, it, vi } from 'vitest';
import { nextTick } from 'vue';
import { createTestingPinia } from '@pinia/testing';
import { useUserStore } from '../../stores/user';

import ChangePassword from '../ChangePassword.vue';

let wrapper;
let user;

describe('Change password page', () => {
	beforeEach(() => {
		wrapper = mount(ChangePassword, {
			global: {
				plugins: [createTestingPinia()],
			},
		});
		user = useUserStore();
	});

	it('Calls the change password method with valid data', async () => {
		const password = wrapper.find('[data-test="password"] input');
		const confirmPassword = wrapper.find(
			'[data-test="confirm-password"] input',
		);
		const form = wrapper.find('[data-test="change-password-form"]');

		expect(wrapper.html()).toMatchSnapshot();

		await password.setValue('Password1!');
		await confirmPassword.setValue('Password1!');

		form.trigger('submit');
		await flushPromises();

		expect(user.changePassword).toHaveBeenCalledOnce();
	});

	it('Renders error message with mismatched passwords when trying to sign up and re-validates form', async () => {
		const password = wrapper.find('[data-test="password"] input');
		const confirmPassword = wrapper.find(
			'[data-test="confirm-password"] input',
		);
		const form = wrapper.find('[data-test="change-password-form"]');

		await password.setValue('Password1!');
		await confirmPassword.setValue('Password41234!');

		await form.trigger('submit');
		await nextTick();
		await flushPromises();

		const error = form.find('.pdap-form-error-message');

		expect(error.exists()).toBe(true);
		expect(error.text()).toBe('Passwords do not match, please try again.');

		await nextTick();

		await confirmPassword.setValue('Pasdasdfasdf');
		await nextTick();
		expect(error.exists()).toBe(true);

		expect(wrapper.html()).toMatchSnapshot();
	});

	it('Handles API error', async () => {
		const mockError = new Error('foo');
		vi.mocked(user.changePassword).mockRejectedValueOnce(mockError);

		const password = wrapper.find('[data-test="password"] input');
		const confirmPassword = wrapper.find(
			'[data-test="confirm-password"] input',
		);
		const form = wrapper.find('[data-test="change-password-form"]');

		await password.setValue('Password1!');
		await confirmPassword.setValue('Password1!');

		form.trigger('submit');
		await flushPromises();

		const error = form.find('.pdap-form-error-message');
		expect(error.exists()).toBe(true);
		expect(error.text()).toBe('foo');
	});
});
