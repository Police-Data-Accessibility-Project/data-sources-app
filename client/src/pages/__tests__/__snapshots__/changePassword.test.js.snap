// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Change password page > Calls the change password method with valid data 1`] = `
<main class="pdap-flex-container mx-auto max-w-2xl">
  <h1>Change your password</h1>
  <form class="pdap-form flex flex-col" id="change-password" name="change-password">
    <!---->
    <div class="pdap-grid-container pdap-input">
      <!---->
      <input id="password" name="password" placeholder="Your updated password" type="password">
      <!---->
      <label class="pdap-input-label" for="password">Password</label>
    </div>
    <div class="pdap-grid-container pdap-input">
      <!---->
      <input id="confirmPassword" name="confirmPassword" placeholder="Confirm your updated password" type="password">
      <!---->
      <label class="pdap-input-label" for="confirmPassword">Confirm Password</label>
    </div>
    <button class="pdap-button pdap-button-primary max-w-full" type="submit">Change password</button>
  </form>
</main>
`;

exports[`Change password page > Renders error message with mismatched passwords when trying to sign up and re-validates form 1`] = `
<main class="pdap-flex-container mx-auto max-w-2xl">
  <h1>Change your password</h1>
  <form class="pdap-form flex flex-col" id="change-password" name="change-password">
    <div class="pdap-form-error-message">Passwords do not match, please try again.</div>
    <div class="pdap-grid-container pdap-input">
      <!---->
      <input id="password" name="password" placeholder="Your updated password" type="password">
      <!---->
      <label class="pdap-input-label" for="password">Password</label>
    </div>
    <div class="pdap-grid-container pdap-input">
      <!---->
      <input id="confirmPassword" name="confirmPassword" placeholder="Confirm your updated password" type="password">
      <!---->
      <label class="pdap-input-label" for="confirmPassword">Confirm Password</label>
    </div>
    <button class="pdap-button pdap-button-primary max-w-full" type="submit">Change password</button>
  </form>
</main>
`;
